PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/secure/IdeaSecureStore.java

```diff
@@ -26,31 +31,49 @@ public static IdeaSecureStore getInstance() {
     }
 
     @Override
-    public void savePassword(@NotNull String serviceName, @NotNull String userName, @Nullable String password) {
-        passwordSafe.setPassword(new CredentialAttributes(serviceName, userName), password);
+    public void savePassword(@Nonnull String serviceName, @Nullable String key, @Nullable String userName, @Nonnull String password) {
+        passwordSafe.setPassword(makeKey(serviceName, key, userName), password);
     }
 
     @Override
     @Nullable
-    public String loadPassword(@NotNull String serviceName, @NotNull String userName) {
-        return passwordSafe.getPassword(new CredentialAttributes(serviceName, userName));
+    public String loadPassword(@Nonnull String serviceName, @Nullable String key, @Nullable String userName) {
+        return passwordSafe.getPassword(makeKey(serviceName, key, userName));
     }
 
-    public void savePassword(@NotNull String key, @NotNull String password) {
-        passwordSafe.setPassword(new CredentialAttributes(key), password);
-    }
-
-    public String loadPassword(@NotNull String key) {
-        return passwordSafe.getPassword(new CredentialAttributes(key));
+    @Override
+    public void forgetPassword(@Nonnull String serviceName, @Nullable String key, @Nullable String userName) {
+        CredentialAttributes oldKey = StringUtils.isNotBlank(userName) ? new CredentialAttributes(key, userName) :
+            new CredentialAttributes(key);
+        passwordSafe.setPassword(oldKey, null);
+        passwordSafe.setPassword(makeKey(serviceName, key, userName), null);
     }
 
     @Override
-    public void forgetPassword(@NotNull String serviceName, @NotNull String userName) {
-        passwordSafe.setPassword(new CredentialAttributes(serviceName, userName), null);
+    public void migratePassword(@Nonnull String oldKeyOrServiceName, @Nullable String oldUsername,
+                                @Nonnull String serviceName, @Nullable String key, @Nullable String userName) {
+        CredentialAttributes oldKey = StringUtils.isNotBlank(oldUsername) ? new CredentialAttributes(oldKeyOrServiceName, userName) :
```

Comment:
I guess: 
oldUsername and username should always be equal and same.
so does overKeyOrServiceName / keyOrServiceName.
right ?

Reply:
I would suggest to keep all older keys and all new keys to make it more easy to under stand.